lua/packer/snapshot.lua

```diff
@@ -0,0 +1,192 @@
+        end
+      end
+
+      res = vim.loop.fs_readdir(dir)
+    end
+  end
+
+  vim.loop.fs_closedir(dir)
+  return completion_list
+end
+
+--- Completion for listing single plugins before taking snapshot
+--- Intended to provide completion for PackerSnapshot command
+snapshot.completion.create = function (lead, cmdline, pos)
+  local cmd_args = (vim.fn.split(cmdline, " "))
+
+  if #cmd_args > 1 then
+    return plugin_complete(lead, cmdline, pos)
+  end
+
+  return {}
+end
+
+--- Completion for listing snapshots in `config.snapshot_path` and single plugins after
+--- the first argument is provided
+--- Intended to provide completion for PackerRollback command
+snapshot.completion.rollback = function (lead, cmdline, pos)
+  local cmd_args = vim.split(cmdline, " ")
+
+  if #cmd_args > 2 then
+    return plugin_complete(lead)
+  else
+    return snapshot.completion.snapshot(lead, cmdline, pos)
+  end
+end
+
+---Serializes a table of git-plugins with `short_name` as table key and another
+---table with `commit`; the serialized tables will be written in the path `snapshot_path`
+---provided, if there is already a snapshot it will be overwritten
+---Snapshotting work only with `plugin_utils.git_plugin_type` type of plugins,
+---other will be ignored.
+---@param snapshot_path string realpath for snapshot file
+---@param plugins table<string, any>[]
+snapshot.create = function(snapshot_path, plugins)
+  assert(type(snapshot_path) == "string",
+    fmt("filename needs to be a string but '%s' provided", type(snapshot_path)))
+  assert(type(plugins) == "table",
+    fmt("plugins needs to be an array but '%s' provided", type(plugins)))
+
+  return async(function()
+    local snapshot_plugins = {}
+    local installed = {}
+    local opt, start = plugin_utils.list_installed_plugins()
+
+    for key, _ in pairs(opt) do installed[key] = key end
+    for key, _ in pairs(start) do installed[key] = key end
+
+    for _, plugin in pairs(plugins) do
+      if installed[plugin.install_path] then -- this plugin is installed
+        log.debug(fmt("Snapshotting '%s'", plugin.short_name))
+        if plugin.type == plugin_utils.git_plugin_type then
+          local rev = await(plugin.get_rev())
+
+          if rev.err then
+            local warn = fmt("Snapshotting %s failed because of error '%s'",
+              plugin.short_name,
+              rev.err
+            )
+            log.warn(warn)
+          else
+            snapshot_plugins[plugin.short_name] = {commit = rev.ok}
+          end
+        end
+      end
+    end
+
+    ---@type string
+    local snapshot_content = "return " .. vim.inspect(snapshot_plugins)
+
+    local fd = loop.fs_open(snapshot_path, "w", tonumber("600", 8))
+
+    if fd == nil then
+      local warn = fmt("Error on creation of snapshot '%s'",snapshot_path)
+      log.warn(warn)
+    else
+      local res = loop.fs_write(fd, snapshot_content)
+      loop.fs_close(fd)
+      if res ~= #snapshot_content then
```

Comment:
Packer should also notify user when snapshot succeeds.

Reply:
I still wonder what should I use to notify the user.
Should I use `log.info` or `print` ?

Reply:
> Should I use log.info or print

I think you should use `vim.notify()` . That's what `PackerCompile` uses .